apps/rush/src/utilities/InstallManager.ts

```diff
@@ -384,9 +382,79 @@ export default class InstallManager {
         }
       }
 
-      // Don't update the file timestamp unless the content has changed, since "rush install"
-      // will consider this timestamp
-      JsonFile.saveJsonFile(tempPackageJson, tempPackageJsonFilename, { onlyIfChanged: true });
+      // NPM expects the root of the tarball to have a directory called 'package'
+      const npmPackageFolder: string = 'package';
+
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2.new"
+      const tempProjectFolder: string = path.join(
+        path.dirname(rushProject.tempPackageTarballFilename),
+        unscopedTempProjectName + '.new');
+
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2\package.json"
+      const tempPackageJsonFilename: string = path.join(tempProjectFolder, RushConstants.packageJsonFilename);
+
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2.gzip"
+      const tarballFile: string = rushProject.tempPackageTarballFilename;
+
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2.old"
+      const extractedFolder: string = tempProjectFolder + '.old';
+
+      // we only want to overwrite the package if the existing tarball's package.json is different from tempPackageJson
+      let shouldOverwrite: boolean = true;
+      try {
+        // extract the tarball and compare the package.json directly
+        if (fsx.existsSync(tarballFile)) {
+
+          // ensure the folder we are about to extract into is clean
+          fsx.removeSync(extractedFolder);
+          fsx.mkdirpSync(extractedFolder);
+
+          tar.extract({
+            cwd: extractedFolder,
+            file: tarballFile,
+            sync: true
+          });
+
+          const extractedPackageJsonFilename: string =
+            path.join(extractedFolder, npmPackageFolder, RushConstants.packageJsonFilename);
+
+          // compare the extracted package.json with the one we are about to write
+          const oldBuffer: Buffer = fsx.readFileSync(extractedPackageJsonFilename);
+          const newBuffer: Buffer = new Buffer(JsonFile.normalize(tempPackageJson));
+
+          if (Buffer.compare(oldBuffer, newBuffer) === 0) {
```

Comment:
I borrowed this code from JsonFile, which you wrote. I suggest we fix this in all places

---
In reply to: [133606562](https://github.com/Microsoft/web-build-tools/pull/324#discussion_r133606562) [](ancestors = 133606562)